src/vs/platform/contextkey/common/scanner.ts

```diff
@@ -0,0 +1,387 @@
+
+				if (lexeme && lexeme.length > 1 && lexeme.startsWith(`'`)) {
+					return `Unexpected token '${token.lexeme}' at offset ${token.offset}. Did you forget to close the string?`;
+				}
+
+				if (lexeme && lexeme.length > 1 && lexeme.endsWith(`'`)) {
+					return `Unexpected token '${token.lexeme}' at offset ${token.offset}. Did you forget to open the string?`;
+				}
+
+				return `Unexpected token '${token.lexeme}' at offset ${token.offset}`;
+			}
+		}
+	}
+
+	private static _whitespace = /^\s*/;
+
+	private static _regexFlags = new Set(['i', 'g', 's', 'm', 'y', 'u'].map(ch => ch.charCodeAt(0)));
+
+	private static _keywords = new Map<string, TokenType>([
+		['not', TokenType.Not],
+		['in', TokenType.In],
+		['false', TokenType.False],
+		['true', TokenType.True],
+	]);
+
+	private _input: string = '';
+	private _start: number = 0;
+	private _current: number = 0;
+	private _errorTokens: Token[] = [];
+
+	get errorTokens(): Readonly<Token[]> {
+		return this._errorTokens;
+	}
+
+	reset(value: string) {
+		this._input = value;
+
+		this._start = 0;
+		this._current = 0;
+		this._errorTokens = [];
+
+		return this;
+	}
+
+	next(): Token {
+		this._eatWhitespace();
+
+		if (this._isAtEnd()) {
+			return { type: TokenType.EOF, offset: this._current };
+		}
+
+		this._start = this._current;
+
+		const ch = this._advance();
+		switch (ch) {
+			case '(': return this._token(TokenType.LParen);
+			case ')': return this._token(TokenType.RParen);
+
+			case '!': {
+				if (this._match('=')) { // support `!=`
+					return this._token(TokenType.NotEq);
+				}
+				return this._token(TokenType.Neg);
+			}
+
+			case '\'': return this._quotedString();
+			case '/': return this._regex();
+
+			case '=':
+				if (this._match('=')) { // support `==`
+					return this._token(TokenType.Eq);
+				} else if (this._match('~')) {
+					return this._token(TokenType.RegexOp);
+				} else {
+					return this._error();
+				}
+
+			case '<': return this._token(this._match('=') ? TokenType.LtEq : TokenType.Lt);
+
+			case '>': return this._token(this._match('=') ? TokenType.GtEq : TokenType.Gt);
+
+			case '&':
+				if (this._match('&')) {
+					return this._token(TokenType.And);
+				} else {
+					return this._error();
+				}
+
+			case '|':
+				if (this._match('|')) {
+					return this._token(TokenType.Or);
+				} else {
+					return this._error();
+				}
+
+			// handle whitespace
+			case ' ':
+			case '\r':
+			case '\t':
+			case '\n': // TODO@ulugbekna: if we're allowing newlines, we should keep track of line # as well ?
+				return this.next();
+
+			case '\u00A0': // &nbsp
+				return this.next();
+
+			default:
+				return this._string();
+		}
+	}
+
+	private _match(expected: string): boolean {
+		if (this._isAtEnd()) {
+			return false;
+		}
+		if (this._input[this._current] !== expected) {
+			return false;
+		}
+		this._current++;
+		return true;
+	}
+
+	private _advance(): string {
+		return this._input[this._current++];
```

Comment:
It would be better to use `String.charCodeAt` instead of the array access and to work everywhere with the `CharCode` const enum. That would be also a low hanging perf improvement.